File: Data Structures & Algorithms/two-integer-sum/submission-11.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mp;
        for(int i = 0; i < nums.size(); i++){
            mp[nums[i]] = i;
        }
        for(int i = 0; i < nums.size(); i++){
            int newtar = target - nums[i];
            if(mp.find(newtar) != mp.end() && i != mp[newtar]){
                return vector<int>{i, mp[newtar]};
            }
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-11.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        curr = []
        nums.sort()
        def dfs(j, tar):
            if tar == 0:
                res.append(curr.copy())
                return
            for i in range(j, len(nums)):
                if nums[i] > tar:
                    break

                curr.append(nums[i])
                dfs(i, tar-nums[i])
                curr.pop()
        dfs(0, target)
        return res

